### PR TITLE
feat: implement ECE metric and Temperature Scaling confidence calibration (#215)

### LINKED ISSUE #215: [Feature]: Model confidence calibration - ECE metric + temperature scaling
## Problem

NightmareNet's evaluation reports accuracy and robustness (correct predictions under distortion) but does NOT assess whether the model's confidence is reliable. A model that outputs 95% confidence on a wrong answer under adversarial attack is more dangerous than one that says "I'm uncertain" - because downstream systems trust high-confidence predictions.

Current evaluation gap:
- No measure of whether confidence correlates with correctness (calibration)
- No detection of when confidence drops under distortion (uncertainty degradation)
- No option to abstain when the model is uncertain (selective prediction)

Paper Section 3.2 mentions the EU AI Act Article 15 requirement for "appropriate level of accuracy, robustness, and cybersecurity" - calibration is a core component of reliability that regulators increasingly expect.

**Relevant files:**
- `nightmarenet/evaluation/evaluator.py` (extends `evaluate()` with calibration metrics)
- `nightmarenet/evaluation/metrics.py` (new metric functions)
- `configs/default.yaml` (add `evaluation.calibration` config block)

### Phase 1: ECE metric + Temperature Scaling (this issue)

1. **`nightmarenet/evaluation/calibration.py`** - New module:
   - `compute_ece(confidences, predictions, labels, n_bins=15)` - Expected Calibration Error
   - `TemperatureScaler.fit(logits, labels)` - Learn optimal temperature T on validation set
   - `TemperatureScaler.calibrate(logits)` - Apply learned temperature to soften/sharpen predictions
   - `reliability_diagram_data(confidences, labels, n_bins)` - Bin data for plotting

2. **Integration with Evaluator:**
   - When `evaluation.calibration.enabled: true`, compute ECE before and after temperature scaling
   - Report in comparison dict: `ece_before`, `ece_after`, `optimal_temperature`
   - Include in markdown report

3. **Config:**
   ```yaml
   evaluation:
     calibration:
       enabled: false
       ece_bins: 15
       temperature_scaling: true
   ```

### Phase 2: Abstention / Selective Prediction (follow-up issue)

Selective prediction (abstain when confidence < threshold) is a separate concern that builds on calibration. Will open as a follow-up after Phase 1 merges.

## Acceptance Criteria

- [ ] `nightmarenet/evaluation/calibration.py` exists with `compute_ece()` and `TemperatureScaler`
- [ ] ECE formula correct: weighted average of |accuracy - confidence| across bins
- [ ] Temperature scaling: single scalar T, learned via NLL minimization on validation logits
- [ ] ECE computed on classification models (requires softmax probabilities, not logits)
- [ ] Integration: `Evaluator.evaluate()` includes calibration metrics when enabled
- [ ] Config: `evaluation.calibration.enabled` (default false, opt-in)
- [ ] Reliability diagram data exposed in JSON (bin_confidences, bin_accuracies, bin_counts)
- [ ] `nightmarenet evaluate` CLI output includes ECE when enabled
- [ ] Test: perfectly calibrated model → ECE ≈ 0
- [ ] Test: overconfident model → ECE > 0, temperature T > 1
- [ ] Test: temperature scaling reduces ECE
- [ ] No regression when `calibration.enabled: false` (default path unchanged)

## Technical Notes

- **ECE is computed on probabilities (after softmax), not logits.** Common implementation bug.
- **Temperature T is a single scalar** optimized via `torch.optim.LBFGS` on the NLL loss over the validation set. Converges in < 1 second for typical datasets.
- **Calibration set must be SEPARATE from test set.** Use the existing `eval_split_ratio` mechanism (the held-out evaluation split) as the calibration set.
- **Reference:** Guo et al. 2017 "On Calibration of Modern Neural Networks" (all deep nets are miscalibrated; temperature scaling is the simplest effective fix).

## Scope

Medium-Large. ~4-5 hours. ECE implementation + temperature scaling + evaluator integration + tests.

---

> **Important:** Please read [CONTRIBUTING.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/CONTRIBUTING.md) in full before requesting assignment. Assignment requests without a technical approach will not be considered. PRs demonstrating thorough research, benchmarks, or creative approaches beyond minimum criteria may earn bonus XP labels.

### PR DESCRIPTION
**Pull request by Aryanbuha890**
### Description
This PR resolves #215 by implementing model confidence calibration metrics (Expected Calibration Error) and temperature scaling via post-softmax probability softening to ensure reliable model predictions in adversarial and out-of-distribution scenarios (complying with standards like the EU AI Act Article 15).

#### 1. Calibration Component
- **[calibration.py](file:///e:/ECSOC-26/NightmareNet/nightmarenet/evaluation/calibration.py)**:
  - Added `compute_ece()` to calculate the Expected Calibration Error grouped into $M$ equally-spaced bins.
  - Added `reliability_diagram_data()` to generate JSON-serializable bin confidences, accuracies, and counts for diagrams.
  - Added `TemperatureScaler` which optimizes a single temperature parameters $T > 0$ using L-BFGS optimization (`torch.optim.LBFGS`) by minimizing Cross-Entropy Loss on validation logits.

#### 2. Evaluator & CLI Integration
- **[evaluator.py](file:///e:/ECSOC-26/NightmareNet/nightmarenet/evaluation/evaluator.py)**:
  - Integrated the calibration pipeline inside `Evaluator.evaluate()`.
  - Splits dataset predictions into calibration split (50% validation) and test split (50%).
  - Fits the temperature parameter, computes ECE before/after calibration, and captures reliability diagram datasets.
  - Added comparison and formatted markdown report outputs.
- **[cli.py](file:///e:/ECSOC-26/NightmareNet/nightmarenet/cli.py)**:
  - Added `--config` and `--checkpoint` arguments to the `evaluate` subcommand.
  - Integrated full dataset evaluation in the `evaluate` CLI handler to print calibration metrics.

#### 3. Configurations & Tests
- **[default.yaml](file:///e:/ECSOC-26/NightmareNet/configs/default.yaml)**: Registered configuration options under `evaluation.calibration`.
- **[test_calibration.py](file:///e:/ECSOC-26/NightmareNet/tests/test_calibration.py)**: Added unit tests verifying perfect calibration (ECE $\approx 0$), overconfident outputs, temperature optimization boundary scaling, and validation ECE reduction.

### Verification Results
- All unit tests completed successfully:
  `tests/test_calibration.py PASSED [100%]`
- Python strict type checking (`mypy`) passed successfully:
  `Success: no issues found in 104 source files`
- Code styling (`ruff`) formatting completed cleanly with no errors.

## Summary by CodeRabbit

* **New Features**
  * Added optional model confidence calibration during evaluation (including Expected Calibration Error before/after and the learned optimal temperature).
  * Added reliability diagram data to support confidence/accuracy analysis.
  * Introduced configuration options to enable calibration, set ECE histogram bins, and toggle temperature scaling.
  * Extended the evaluation CLI with config-driven evaluation, checkpoint/model selection, and direct JSON output.
* **Bug Fixes**
  * Improved validation for required evaluation inputs and ensured device selection follows the selected execution device.
* **Tests**
  * Added coverage for calibration utilities and calibration behavior when temperature scaling is disabled.

### CODE REVIEW

**Comment by Adit-Jain-srm** [comment]
CI green (Python 3.12 = known OpenAPI drift). Partial AC gap.

### Fix before merge

- [ ] **`evaluator.py` ~L465**: `temperature_scaling` config flag is ignored. Code always runs temperature scaling when calibration is enabled, regardless of the `temperature_scaling: true/false` setting. Gate it:
  ```python
  if cal_cfg.get("temperature_scaling", True):
      # run temperature scaling
  ```
- [ ] **`cli.py` ~L185**: JSON output with numpy values will crash `json.dumps`. Add a numpy-safe encoder or use `default=str`.
- [ ] **Star/follow not completed.** Pre-submission checklist requires starring the repo and following @Adit-Jain-srm.

### Advisory (non-blocking)

1. `evaluator.py` L411: `raise e` should be `raise ... from e` to preserve traceback context.

<details>
<summary>Prompt for AI Agents</summary>

Verify each finding against current code. Fix only still-valid issues, skip the rest with a brief reason, keep changes minimal, and validate.

1. In `nightmarenet/evaluation/evaluator.py` around L465, gate temperature scaling on config:
```python
if cal_cfg.get("temperature_scaling", True):
    scaler = TemperatureScaler()
    scaler.fit(logits, labels)
    results["calibration"]["temperature"] = scaler.temperature
```

2. In `nightmarenet/cli.py` ~L185, add numpy-safe JSON serialization:
```python
class _NumpyEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, (np.integer,)): return int(obj)
        if isinstance(obj, (np.floating,)): return float(obj)
        if isinstance(obj, np.ndarray): return obj.tolist()
        return super().default(obj)

# Use: json.dumps(results, cls=_NumpyEncoder)
```

3. In `evaluator.py` L411: `raise e` -> `raise RuntimeError("Calibration failed") from e`

4. Star the repo and follow @Adit-Jain-srm.

Commit message: `fix: gate temperature_scaling on config, add numpy JSON encoder`
</details>

**Review by Adit-Jain-srm — CHANGES_REQUESTED** [review]
Please implement these before merge @Aryanbuha890

**Comment by Aryanbuha890** [comment]
@Adit-Jain-srm Please review this PR also.

**Comment by Aryanbuha890** [comment]
> @Adit-Jain-srm Please review this PR also.

Please now verify this.

**Comment by Aryanbuha890** [comment]
@Adit-Jain-srm Please check this and merge this.

**Review by Adit-Jain-srm — APPROVED** [review]
LGTM
